crates/ecs/src/lib.rs

```diff
@@ -585,27 +871,51 @@ impl<'a, Component> Deref for Read<'a, Component> {
     }
 }
 
-impl<Component: Send + Sync + 'static + Sized> SystemParameter for Read<'_, Component> {
-    type BorrowedData<'components> = ReadComponentVec<'components, Component>;
+#[inline(always)]
+fn get_and_inc(value: &mut usize) -> usize {
+    let tmp = *value;
+    *value += 1;
+    tmp
+}
 
-    fn borrow(world: &World) -> Self::BorrowedData<'_> {
-        world.borrow_component_vec::<Component>()
+impl<Component: Debug + Send + Sync + 'static + Sized> SystemParameter for Read<'_, Component> {
+    type BorrowedData<'components> = (
+        usize,
+        Vec<(usize, ReadComponentVec<'components, Component>)>,
+    );
+
+    fn borrow<'a>(world: &'a World, signature: &[TypeId]) -> Self::BorrowedData<'a> {
+        (
+            0,
+            world
+                .borrow_component_vecs_with_signature::<Component>(signature)
+                .into_iter()
+                .map(|component_vec| (0, component_vec))
+                .collect(),
+        )
     }
 
-    unsafe fn fetch_parameter(
-        borrowed: &mut Self::BorrowedData<'_>,
-        entity: Entity,
-    ) -> Option<Self> {
-        if let Some(component_vec) = borrowed {
-            if let Some(Some(component)) = component_vec.get(entity.id) {
-                return Some(Self {
-                    // The caller is responsible to only use the
-                    // returned value when BorrowedData is still in scope.
-                    #[allow(trivial_casts)]
-                    output: &*(component as *const Component),
-                });
+    unsafe fn fetch_parameter(borrowed: &mut Self::BorrowedData<'_>) -> Option<Option<Self>> {
+        if let Some(archetype) = (borrowed.1).get_mut(borrowed.0) {
+            if let Some(component_vec) = &archetype.1 {
+                return if let Some(component) = component_vec.get(get_and_inc(&mut archetype.0)) {
+                    if let Some(component) = component {
+                        return Some(Some(Self {
+                            // The caller is responsible to only use the
+                            // returned value when BorrowedData is still in scope.
+                            #[allow(trivial_casts)]
+                            output: &*(component as *const Component),
+                        }));
+                    }
+                    Some(None)
```

Comment:
this is a code smell. If you're working with multiple `Option`s then you should chain them together using combinators, not wrap them inside of each other!